### PR TITLE
Adding support for the BMA400 accelerometer

### PR DESCRIPTION
**Pull request by BST-Github-Admin**
Hi,

Adding support for the BMA400 accelerometer.

### CODE REVIEW

File: drivers/sensor/bma400/Kconfig [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# BMA400 3 Axis Accelerometer configuration
+
+# Copyright (c) 2020 Bosch Sensortec GmbH
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig BMA400
```
Comment (galak) [inline drivers/sensor/bma400/Kconfig:6]: this should be `config BMA400`

File: drivers/sensor/bma400/bma400.c [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+				&data[i], 1);
+		}
+	} else   {
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+static void delay_us(u32_t period)
```
Comment (galak) [inline drivers/sensor/bma400/bma400.c:186]: remove delay_us and just use k_usleep directly.

**Review by galak — COMMENTED** [review]
Devicetree bits look fine.  A few comments.

**Comment by BoschSensortec** [comment]
Hi,
Since Chocolatey is currently not available it is not possible to test on Windows. Kindly bear with the ping-pong between the CI test and related fixes.

File: drivers/sensor/bma400/Kconfig (hunk 1 above)
Comment (BoschSensortec) [reply drivers/sensor/bma400/Kconfig]: changed from menuconfig to config

File: drivers/sensor/bma400/bma400.c (hunk 2 above)
Comment (BoschSensortec) [reply drivers/sensor/bma400/bma400.c]: removed

**Comment by mbolivar-nordic** [comment]
> Devicetree bits look fine. A few comments.

No chance we can make this multi-instance like we already did for BME280? https://github.com/zephyrproject-rtos/zephyr/pull/23741

**Comment by BoschSensortec** [comment]
> > Devicetree bits look fine. A few comments.
> 
> No chance we can make this multi-instance like we already did for BME280? #23741

Do you mean SPI and I2C support?

**Comment by mbolivar-nordic** [comment]
> > No chance we can make this multi-instance like we already did for BME280? #23741
> 
> Do you mean SPI and I2C support?

That would be multiple bus support. "Multi-instance" means automatic allocation of device instances based on the contents of the devicetree using DT_INST_FOREACH as done in #23741.

**Comment by mbolivar-nordic** [comment]
See also https://docs.zephyrproject.org/latest/guides/dts/howtos.html#create-struct-devices-in-a-driver

**Comment by BoschSensortec** [comment]
> See also https://docs.zephyrproject.org/latest/guides/dts/howtos.html#create-struct-devices-in-a-driver

Probably in an upcoming update. This PR was meant only for basic functionality. The upcoming update will also include interrupts and features like the Step counter.

**Comment by mbolivar-nordic** [comment]
> This PR was meant only for basic functionality.

OK, understood. No objections.

File: drivers/sensor/bma400/bma400.h [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+#ifndef ZEPHYR_DRIVERS_SENSOR_BMA400_BMA400_H_
+#define ZEPHYR_DRIVERS_SENSOR_BMA400_BMA400_H_
+
+#include <device.h>
+#include <sys/util.h>
+#include <zephyr/types.h>
+#include <drivers/sensor.h>
+
+#define BMA400_REG_CHIP_ID          (0x00)
```
Comment (jfischer-phytec-iot) [inline drivers/sensor/bma400/bma400.h:15]: Why there are parentheses for these single tokens? https://www.kernel.org/doc/html/v4.10/process/coding-style.html#macros-enums-and-rtl

File: drivers/sensor/bma400/bma400.c [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+static void channel_accel_convert(struct sensor_value *val, s64_t raw_val,
+	u8_t range);
+static int set_accel_config(const struct bma400_acc_cfg *cfg,
+	struct bma400_dev *dev);
+static s16_t convert_acc_data(u8_t lsb, u8_t msb);
+static int set_power_mode(u8_t power_mode, struct bma400_dev *dev);
+static int set_attributes(double odr, double osr, double range,
+	struct bma400_dev *dev);
+
+int bma400_init(struct device *dev)
```
Comment (jfischer-phytec-iot) [inline drivers/sensor/bma400/bma400.c:33]: Typically init is on the bottom.

File: drivers/sensor/bma400/bma400.c [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+#include <init.h>
+#include <drivers/sensor.h>
+#include <sys/__assert.h>
+#include <logging/log.h>
+
+#include "bma400.h"
+
+LOG_MODULE_REGISTER(bma400, CONFIG_SENSOR_LOG_LEVEL);
+
+static int reg_read(u8_t reg, u8_t *data, u16_t length,
```
Comment (jfischer-phytec-iot) [inline drivers/sensor/bma400/bma400.c:19]: These forward declarations are not necessary, remove it / reorganize your code

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: drivers/sensor/bma400/bma400.c (hunk 4 above)
Comment (BoschSensortec) [reply drivers/sensor/bma400/bma400.c:33]: @jfischer-phytec-iot , This is the preferred style and would prefer keeping it that way.

Comment (jfischer-phytec-iot) [reply drivers/sensor/bma400/bma400.c:33]: It is not, the most of the driver have init on the bottom. Please fix this style issues.

File: samples/sensor/bma400/boards/nrf52840dk_nrf52840.overlay [hunk 6]
```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright (c) 2020 Bosch Sensortec GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&i2c0 {
```
Comment (jfischer-phytec-iot) [inline samples/sensor/bma400/boards/nrf52840dk_nrf52840.overlay:7]: You can make it much more generic. As there is not any board in the tree equipped this device, you can rename this overlay to app.overlay, move it to  _samples/sensor/bma400/_ , and rename i2c0 to arduino_i2c. With this changes one can use this sample on any board which provides a configuration for Arduino connectors.

```diff
diff --git a/samples/sensor/bma400/boards/nrf52840dk_nrf52840.overlay b/samples/sensor/bma400/boards/nrf52840dk_nrf52840.overlay
index 10bf917c5a..b28abb5995 100644
--- a/samples/sensor/bma400/boards/nrf52840dk_nrf52840.overlay
+++ b/samples/sensor/bma400/boards/nrf52840dk_nrf52840.overlay
@@ -4,7 +4,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-&i2c0 {
+&arduino_i2c {
        bma400@14 {
                compatible = "bosch,bma400";
                reg = <0x14>;
```

**Review by jfischer-no — CHANGES_REQUESTED** [review]

**Comment by mbolivar-nordic** [comment]
@BoschSensortec any plans to update this PR?

**Pull request closed without merging**